### DISCUSSION TITLE
APIRouter testing with TestClient always return assertion error 404

### DISCUSSION THREAD

**Discussion by mydevlocal** [fastapi/fastapi#7582] [Questions] [answered]
### Description

I'm about to create restapi and add testing files for each route. I'm using `APIRouter` for all my path operations, but when i create a test file, i always get assertion error 404. If i change the path operation using `FastAPI`, the test passed. Did i miss something maybe like specific configuration for testing path operations from `APIRouter` instance?

Here is the example of my project structure:

```
app/
|__routes/
|____basic.py
|__tests/
|____test_basic.py
|__server.py
|__main.py
```

`basi.py`
```
from fastapi import APIRouter

router = APIRouter()

@router.get('/')
def read_root():
    return {"Hello": "World"}
```
`main.py`
```
from fastapi import FastAPI

app: FastAPI = FastAPI()
```

`server.py`
```
from app.main import app
from app.routes.basic import router as basic_router
[...]
ROUTERS = (basic.router,)

app.include_router(ROUTERS[0], tags=["Basic"])
[...]

if __name__ == "__main__":
    import uvicorn

    uvicorn.run("app.main:app", host=getenv('APP_HOST'),
                port=int(getenv('APP_PORT')), reload=True, log_level="debug")
```

`test_basic.py`
```
from fastapi.testclient import TestClient
from app.main import app

client = TestClient(app)

def test_read_root():
    response = client.get("/")

    assert response.status_code == 200
```

**Comment by phy25** [comment, marked as answer]
Your `test_basic.py` needs `from app.server import app` or you won't get those router included.

**Comment by mydevlocal** [comment]
> Your `test_basic.py` needs `from app.server import app` or you won't get those router included.

OMG, you're right. I set `from app.server import app` to my test files, and it works.
Thank you so much.

**Comment by phy25** [comment]
If this works for you you could close the issue. :)

**Comment by tiangolo** [comment]
Thanks for the help here @phy25 ! :clap: :bow: 

Thanks for reporting back and closing the issue @mydevlocal :+1:
